cf-agent/package_module.c

```diff
@@ -0,0 +1,1611 @@
+            Log(LOG_LEVEL_VERBOSE, "unsupported option: %s", line);
+        }
+    }
+    
+    return package_data;
+}
+
+static int NegotiateSupportedAPIVersion(PackageModuleWrapper *wrapper)
+{
+    int api_version = -1;
+
+    Rlist *response = NULL;
+    if (ReadWriteDataToPackageScript("supports-api-version", "",
+                                     &response, wrapper) != 0)
+    {
+        Log(LOG_LEVEL_INFO, "Some error occurred while communicating with "
+            "wrapper while getting supported API version.");
+        return -1;
+    }
+    
+    if (response)
+    {
+        if (RlistLen(response) == 1)
+        {
+            api_version = atoi(RlistScalarValue(response));
+            Log(LOG_LEVEL_DEBUG, "package wrapper API version: %d", api_version);
+        }
+        RlistDestroy(response);
+    }
+    return api_version;
+}
+
+/* IMPORTANT: this might not return all the data we need like version
+              or architecture but package name MUST be known. */
+static
+PackageInfo *GetPackageData(const char *name, Rlist *options,
+                            const PackageModuleWrapper *wrapper)
+{   
+    char *options_str = ParseOptions(options);
+    char *request = StringFormat("%sFile=%s\n", options_str, name);
+    
+    Rlist *response = NULL;
+    if (ReadWriteDataToPackageScript("get-package-data", request, &response,
+            wrapper) != 0)
+    {
+        Log(LOG_LEVEL_INFO, "Some error occurred while communicating with "
+            "package module while collecting package data.");
+        free(options_str);
+        free(request);
+        return NULL;
+    }
+    
+    PackageInfo *package_data = NULL;
+    
+    if (response)
+    {   
+        package_data = ParseAndCheckPackageDataReply(response);
+        RlistDestroy(response);
+        
+        if (package_data)
+        {
+            /* At this point at least package name MUST be known (if no error) */
+            if (!package_data->name || package_data->type == PACKAGE_TYPE_NONE)
+            {
+                Log(LOG_LEVEL_INFO, "Unknown package name.");
+                FreePackageInfo(package_data);
+                package_data = NULL;
+            }
+        }
+    }
+    free(options_str);
+    free(request);
+        
+    return package_data;
+}
+
+static char *GetPackageModuleRealPath(const char *package_module_name)
+{
+    
+    return StringFormat("%s%c%s%c%s", GetWorkDir(), FILE_SEPARATOR,
+                        "package_modules", FILE_SEPARATOR,
+                        package_module_name);
+}
+
+static int IsPackageInCache(EvalContext *ctx,
+                            const PackageModuleWrapper *module_wrapper,
+                            const char *name, const char *ver, const char *arch)
+{
+    const char *version = ver;
+    /* Handle latest version in specific way for repo packages. 
+     * Please note that for file packages 'latest' version is not supported
+     * and check against that is made elsewhere. */
```

Comment:
It would be a kindness to the reader to be explicit: "elsewhere" leaves the reader to search through many MB of code for where, exactly, the check is made.